addons/account/models/account_move.py

```diff
@@ -1,347 +1,1738 @@
-    _inherit = ['mail.thread', 'mail.activity.mixin']
 
     @api.model
-    def default_get(self, fields):
-        rec = super(AccountMove, self).default_get(fields)
-        if not rec.get('journal_id'):
-            rec.update({'journal_id': self.env['account.journal'].search([('type', '=', 'general'), ('company_id', '=', self.env.company.id)], limit=1).id})
-        return rec
+    def _get_default_journal(self):
+        ''' Get the default journal.
+        - The default journal could be passed through the context using the 'default_journal_id' key containing its id.
+        - The default journal is determined by the default type found in the context or 'entry'.
+        '''
+        move_type = self._context.get('default_type', 'entry')
+        journal_type = 'general'
+        company_id = self._context.get('default_company_id', self.env.company.id)
+        if move_type in ('out_invoice', 'out_refund', 'out_receipt'):
+            journal_type = 'sale'
+        elif move_type in ('in_invoice', 'in_refund', 'in_receipt'):
+            journal_type = 'purchase'
+
+        if self._context.get('default_journal_id'):
+            journal = self.env['account.journal'].browse(self._context['default_journal_id'])
+
+            if move_type != 'entry' and journal.type != journal_type:
+                raise UserError(_("Cannot create an invoice of type %s with a journal having %s as type.") % (move_type, journal_type))
+        else:
+            domain = [('company_id', '=', company_id), ('type', '=', journal_type)]
+            journal = self.env['account.journal'].search(domain, limit=1)
+
+            if not journal:
+                raise UserError(_('Please define an accounting %s journal in your company.') % journal_type)
+        return journal
+
+    @api.model
+    def _get_default_currency(self):
+        ''' Get the default currency from either the journal or either the company. '''
+        journal = self._get_default_journal()
+        return journal.currency_id or journal.company_id.currency_id
+
+    @api.model
+    def _get_default_invoice_incoterm(self):
+        ''' Get the default incoterm for invoice. '''
+        return self.env.company.incoterm_id
+
+    # ==== Business fields ====
+    name = fields.Char(string='Number', required=True, readonly=True, copy=False, default='/')
+    date = fields.Date(string='Date', required=True, index=True, readonly=True,
+        states={'draft': [('readonly', False)]},
+        default=fields.Date.context_today)
+    ref = fields.Char(string='Reference', copy=False, readonly=True,
+        states={'draft': [('readonly', False)]})
+    narration = fields.Text(string='Internal Note')
+    state = fields.Selection(selection=[
+            ('draft', 'Unposted'),
+            ('posted', 'Posted'),
+            ('cancel', 'cancelled')
+        ], string='Status', required=True, readonly=True, copy=False, tracking=True,
+        default='draft')
+    type = fields.Selection(selection=[
```

Comment:
Use `selection_add` for that.

Reply:
Can't we use `selection_add` there?

Reply:
@pedrobaeza that just solve add items to list BTW still must redefine:
domain setted on onchange_journal, this type define how move works (debit, credit values) and bases on type journal type is selected.
Use a variable for that will allow just add to that and less code.

Reply:
I don't think so. Overriding methods and checking the value of the field and doing logic according that, or calling super otherwise, is the way to do it. Unless you say an specific problem, in general I don't see any...